### ISSUE TITLE
cap target output

### ISSUE THREAD

**Issue by mattou78400** [creme-ml/creme#248]
Hi all,

Is there any way I could set a range within which the model will predict the result?
e.g.

I have variation data (between -0.2 and +0.2) and I'd like to tell the model, discard anything predicted above or below this range.

Thanks!

**Comment by raphaelsty** [comment]
Hi @mattou78400, 

Maybe you could clip your model's predictions. 

If your model predicts more than 0.2, then you will reduce the prediction to 0.2 and vice versa.

[numpy clipping function](https://docs.scipy.org/doc/numpy/reference/generated/numpy.clip.html) :-)

**Comment by mattou78400** [comment]
Thanks!

But will the model consider it when training?

**Comment by raphaelsty** [comment]
The model will not take this into account during the learning process.

You might also want to add some regularization to your model. 

For example, if you use linear regression, increasing the value of parameter l2 will force the model weights to be close to 0.

The standard scaler is also useful with a linear model to fit well. :-)

If your weights are closer to 0, the scale of your predictions should respect the target scale

**Comment by greatsharma** [comment]
Hey @mattou78400, first of all @raphaelsty indicated you in the right direction and you can try that out if it satisfies your requirements.
But if still you are not satisfied with clipping thing and want to include this implicitly in your model's training then I think you should make a function similar to tanh, just like tanh squashes the real line to a range of -1, 1 your custom made function should squash it to range -0.2, 0.2. You can think of this like the logistic regression in which we map the output to 0, 1 range using sigmoid. The benefits of this is that instead of hard cutting the outputs to a threshold you can now map them which results in maintaining the original relation between the points before and after mapping, eg. both 10 and 100 will be clipped to 0.2 and hence you will loose the essence of how different they were before clipping but  using a squashing function f, f(20) > f(10) and that information of is still maintained. Also because the function is smooth and continuous hence easily differentiable and therefore can be very easily incorporated in your model's training to calculate gradients like stuff.

Also you can use regularisation as indicated by @raphaelsty but because it's a penalization/forcing kind of thing so it will not gaurantee that all your outputs will be in the required range, they may bipass the range.

**Comment by gbolmier** [comment]
Hi @mattou78400! For clipping after prediction we recently implemented `meta.PredClipper` that will be available in the next release. If you don't want to wait, you can use it by installing the dev version or just use `numpy.clip` as pointed by @raphaelsty. Hope it helps :)

**Comment by MaxHalford** [comment]
Most has been said, I'll just give my 2 cents. There are two main ways I would see for dealing with your problem.

The first technique is to filter out your observations that are not in range `{-0.2, +0.2}`:

```python
for x, y in stream:
    if not -0.2 < y < +0.2:
        continue
    model.fit_one(x, y)
```

It's simple, to the point, and ensures that your model will never see observations who's target is in the specified interval. What it doesn't ensure, however, is that your model will never produce predictions outside of the specified interval. Also filtering training data means your model is using less training data and thus might be less powerful. As @gbolmier said, in the next release there will be a `meta.PredClipper` for making sure that predictions are capped.

**Comment by mattou78400** [comment]
Thank you all for your answers

I have a lot of tricks to try out! 
I'll post an update once I figured it all out!

**Comment by MaxHalford** [comment]
Can I close this @mattou78400?
